lib/src/main/java/org/esbtools/eventhandler/ResponsePromise.java

```diff
@@ -0,0 +1,40 @@
+/*
+ *  Copyright 2015 esbtools Contributors and/or its affiliates.
+ *
+ *  This file is part of esbtools.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.esbtools.eventhandler;
+
+/**
+ * An asynchronous capture of potentially yet-to-come responses.
+ *
+ * <p>Allows code to build a {@link Result} object which may populate lazily with results from the
+ * {@link ResponsesHandler} passed to {@link #then(ResponsesHandler)}.
+ * @param <T> The type of requests
+ * @param <U> The type of responses
+ */
+public interface ResponsePromise<T, U> {
```

Comment:
See above note about CompletableFuture, I could maybe make it work, but currently I think it would be more awkward than not. Open to ideas though.